src/objectid.ts

```diff
@@ -59,67 +54,77 @@ export class ObjectId {
   constructor(id?: string | Buffer | number | ObjectIdLike | ObjectId) {
     // Duck-typing to support ObjectId from different npm packages
     if (id instanceof ObjectId) {
-      this.id = id.id;
+      this._id = id.id;
       this.__id = id.__id;
-      return;
+    }
+
+    if (typeof id === 'object' && 'id' in id) {
+      if ('toHexString' in id && typeof id.toHexString === 'function') {
+        this._id = Buffer.from(id.toHexString(), 'hex');
+      } else {
+        this._id = typeof id.id === 'string' ? Buffer.from(id.id) : id.id;
+      }
     }
 
     // The most common use case (blank id, new objectId instance)
     if (id == null || typeof id === 'number') {
       // Generate a new id
-      this.id = ObjectId.generate(typeof id === 'number' ? id : undefined);
+      this._id = ObjectId.generate(typeof id === 'number' ? id : undefined);
       // If we are caching the hex string
-      if (ObjectId.cacheHexString) this.__id = this.toString('hex');
-      // Return the object
-      return;
+      if (ObjectId.cacheHexString) {
+        this.__id = this.id.toString('hex');
+      }
     }
 
-    // Check if the passed in id is valid
-    const valid = ObjectId.isValid(id);
+    if (ArrayBuffer.isView(id) && id.byteLength === 12) {
+      this._id = ensureBuffer(id);
+    }
 
-    // Throw an error if it's not a valid setup
-    if (!valid) {
+    if (typeof id === 'string') {
+      if (id.length === 12) {
+        const bytes = Buffer.from(id);
+        if (bytes.byteLength === 12) {
+          this._id = bytes;
+        }
+      }
+      if (id.length === 24 && checkForHexRegExp.test(id)) {
+        this._id = Buffer.from(id, 'hex');
+      }
+    }
+
+    if (typeof this['_id'] === 'undefined') {
```

Comment:
is this path reachable? `_id` is defined as non-optional

Reply:
This was checking if the string length/format checks failed, I made the control flow clearer.